### ISSUE TITLE
Query<object>.Indices won't Serialize

### ISSUE THREAD

**Issue by abibell** [elasticsearch/elasticsearch-net#1187]
```
var termQuery = Query<object>.Term("field", "valueText");
var indicesQuery = Query<object>.Indices(i => i.Indices(new[] { "test" }));
//indicesQuery won't Serialize
var q = this.SerializeThenDeserialize(
                f => f.Bool,
                f => f.Bool(b => b
                    .Must(termQuery, indicesQuery)
                    .Should(termQuery, indicesQuery))
            );
```
Any ideas on fix options? Happy to help, this would clean our code

**Comment by Mpdreamz** [comment]
Thats because its [conditionless](https://github.com/elasticsearch/elasticsearch-net/blob/develop/src/Nest/DSL/Query/IndicesQueryDescriptor.cs#L57) it needs a query component. The elasticsearch [docs on indices query](http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/query-dsl-indices-query.html) says so as well.

So you'd want to pass the term query like so:

```
.Query(q=>
    q.Indices(i=>i
        .Indices(new [] { "test" })
        .Query(iq=>iq.Term("field", "valueText"))
    )
)
```
Read more about conditionless queries here:

http://nest.azurewebsites.net/nest/writing-queries.html

It also describes how to turn off if you really must :) It also goes into great detail on how you can leverage (`&&`, `||` and `!`) to write terser bool queries. 

Hope this helps!

**Comment by abibell** [comment]
I tried to simplify the problem. Here is the actual test where **should** contains information, but it turns out empty. I think we should reopen as I was trying to simplify the problem. The bigger target is. We are trying to enforce security on certain index-type combination (we can have same type name in different indexes) based on users subscriptions. At the moment we use an ugly Json serialisation and SearchDescriptor raw building, too much code for simple thing.

```csharp
[Test]
	    public void IsClientSideSearchQueryDeserialisable()
	    {
	        var newDescriptor = Deserialize<SearchDescriptor<dynamic>>(original);
            var actual = Serialize(newDescriptor);//Should is empty

            var descriptorJobject = JObject.Parse(actual);
	        var expressionList = new[]
	                             {
	                                 new[] {"size"},//Works
	                                 new[] {"from"},//Works
	                                 new[] {"query", "filtered", "query", "query_string", "fields"},//Works
	                                 new[] {"query", "filtered", "filter", "bool", "must"},//Works
//Can't find contents of should
                                     new[] {"query", "filtered", "filter", "bool", "should", "0", "indices", "filter"},
	                                 new[] {"aggs", "Databases", "terms", "field"},//Works
	                                 new[] {"aggs", "Year", "terms", "field"}//Works
	                             };
	        foreach (var e in expressionList)
	            VerifyJsonPath(descriptorJobject, e);

            //If we do a search without the below line
            //it seems to contect /_all/object/_search instead of /_search
	        newDescriptor.AllTypes().AllIndices();
	    }

        private static void VerifyJsonPath(JToken descriptorJobject, IEnumerable<string> strings)
        {
            foreach (var item in strings)
            {
                Assert.IsNotNull(descriptorJobject[item]);
                descriptorJobject = descriptorJobject[item];
            }
        }

	    public string Serialize<T>(T obj) where T : class
        {
            var json = Encoding.UTF8.GetString(_client.Serializer.Serialize(obj));
            return json;
        }

        public T Deserialize<T>(string json) where T : class
        {
            return _client.Serializer.Deserialize<T>(new MemoryStream(Encoding.UTF8.GetBytes(json)));
        }

#region Our Search query

	        const string original =
                @"{
   ""size"":10,
   ""from"":0,
   ""query"":{
      ""filtered"":{
         ""query"":{
            ""query_string"":{
               ""query"":""*"",
               ""default_operator"":""AND"",
               ""fields"":[
                  ""type1._all"",
                  ""type2.title"",
                  ""type3._all"",
                  ""type4._all""
               ]
            }
         },
         ""filter"":{
            ""bool"":{
               ""must"":[

               ],
               ""should"":[
                  {
                     ""indices"":{
                        ""index"":""index1"",
                        ""filter"":{
                           ""terms"":{
                              ""_type"":[
                                 ""type1"",
                                 ""type2"",
                                 ""type3""
                              ]
                           }
                        },
                        ""no_match_filter"":""none""
                     }
                  },
                  {
                     ""indices"":{
                        ""index"":""index2"",
                        ""filter"":{
                           ""terms"":{
                              ""_type"":[
                                 ""type2"",
                                 ""type4""
                              ]
                           }
                        },
                        ""no_match_filter"":""none""
                     }
                  }
               ]
            }
         }
      }
   },
   ""aggs"":{
      ""Databases"":{
         ""terms"":{
            ""field"":""_type"",
            ""size"":5
         }
      },
      ""Index"":{
         ""terms"":{
            ""field"":""_index"",
            ""size"":5
         }
      },
      ""Year"":{
         ""terms"":{
            ""field"":""publicationYear"",
            ""size"":5
         }
      }
   }
}";

	        #endregion
```

**Comment by Mpdreamz** [comment]
Sorry for closing this too eagerly @abibell ! on closer inspection it seems we do not have the indices filter mapped:

http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/query-dsl-indices-filter.html

Only its query counterpart. I am working on adding this in now and should be available in the next release.

I would also love to help you to get your query building to be all in code feel free to create a ticket with how you have it now and how it could be terser.

**Comment by Mpdreamz** [comment]
Hi @abibell I just pushed a fix that implements the missing indices filter that also support the `none` and `all` shortcuts on the `no_match_filter `. I've also updated the indices query to support the shortcuts (read and write) as well.

See your test here (with some small tweaks here):

https://github.com/elasticsearch/elasticsearch-net/blob/19a580496425c39c1665562de4b5b2b996b3282f/src/Tests/Nest.Tests.Unit/Reproduce/Reproduce1187Tests.cs

That now passes.
